velox/functions/prestosql/tests/FunctionBaseTest.h

```diff
@@ -214,6 +214,119 @@ class FunctionBaseTest : public testing::Test {
     return vectorMaker_.arrayOfRowVector(rowType, data);
   }
 
+  // Create an ArrayVector<ArrayVector<T>> from nested std::vectors of values.
+  // Example:
+  //   std::vector<std::optional<int64_t>> a {1, 2, 3};
+  //   std::vector<std::optional<int64_t>> b {4, 5};
+  //   std::vector<std::optional<int64_t>> c {6, 7, 8};
+  //   auto O = [](const std::vector<std::optional<int64_t>>& data) {
+  //     return std::make_optional(data);
+  //   };
+  //   auto arrayVector = makeNestedArrayVector<int64_t>(
+  //      {{O(a), O(b)},
+  //       {O(a), O(c)},
+  //       {O({})},
+  //       {O({std::nullopt})}});
+  //
+  //   EXPECT_EQ(4, arrayVector->size());
+  template <typename T>
+  ArrayVectorPtr makeNestedArrayVector(
+      const std::vector<
+          std::vector<std::optional<std::vector<std::optional<T>>>>>& data) {
+    vector_size_t size = data.size();
+    BufferPtr offsets = AlignedBuffer::allocate<vector_size_t>(size, pool());
+    BufferPtr sizes = AlignedBuffer::allocate<vector_size_t>(size, pool());
+    BufferPtr nulls = AlignedBuffer::allocate<uint64_t>(size, pool());
+
+    auto rawOffsets = offsets->asMutable<vector_size_t>();
+    auto rawSizes = sizes->asMutable<vector_size_t>();
+    auto rawNulls = nulls->asMutable<uint64_t>();
+
+    // Flatten the outermost layer of std::vector of the input, and populate
+    // the sizes, offsets, and nulls for the top-level array vector.
+    std::vector<std::optional<std::vector<std::optional<T>>>> flattenedData;
+    vector_size_t i = 0;
+    for (const auto& vector : data) {
+      flattenedData.insert(flattenedData.end(), vector.begin(), vector.end());
+
+      rawSizes[i] = vector.size();
+      rawOffsets[i] = (i == 0) ? 0 : rawOffsets[i - 1] + rawSizes[i - 1];
+      bits::setBit(rawNulls, i, true);
```

Comment:
Shouldnt this be set only if size is 0 ?

Reply:
The null buffer uses 1 to represent "not null" and 0 to represent "null" at each row. So here I'm setting "not null" at every row. This is indeed a bit confusing. Plus the null buffer is actually unnecessary because the argument type of this function doesn't allow nulls at the top level, so I'll remove the statements about the null buffer in this function. Thanks for catching this!